File: src/main/java/no/nav/fo/veilarbregistrering/arbeidssoker/Arbeidssokerperioder.java
package no.nav.fo.veilarbregistrering.arbeidssoker;

import no.nav.fo.veilarbregistrering.bruker.Periode;

import java.util.Comparator;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static java.util.Collections.emptyList;
import static java.util.stream.Collectors.toList;
import static no.nav.fo.veilarbregistrering.arbeidssoker.Arbeidssokerperiode.EldsteFoerst.eldsteFoerst;

public class Arbeidssokerperioder {

    private final List<Arbeidssokerperiode> arbeidssokerperioder;

    public static Arbeidssokerperioder of(List<Arbeidssokerperiode> arbeidssokerperioder) {
        return new Arbeidssokerperioder(arbeidssokerperioder);
    }

    public Arbeidssokerperioder(List<Arbeidssokerperiode> arbeidssokerperioder) {
        this.arbeidssokerperioder = arbeidssokerperioder != null ? arbeidssokerperioder : emptyList();
    }

    public List<Arbeidssokerperiode> overlapperMed(Periode forespurtPeriode) {
        return arbeidssokerperioder.stream()
                .filter(p -> p.getPeriode().overlapperMed(forespurtPeriode))
                .filter(p -> p.getFormidlingsgruppe().erArbeidssoker())
                .sorted(Comparator.comparing(e -> e.getPeriode().getFra()))
                .collect(toList());
    }

    public boolean dekkerHele(Periode forespurtPeriode) {
        Optional<Arbeidssokerperiode> eldsteArbeidssokerperiode = arbeidssokerperioder.stream()
                .sorted(Comparator.comparing(e -> e.getPeriode().getFra()))
                .findFirst();

        return eldsteArbeidssokerperiode
                .map(arbeidssokerperiode -> forespurtPeriode.fraOgMed(arbeidssokerperiode.getPeriode()))
                .orElse(false);
    }

    public List<Arbeidssokerperiode> asList() {
        return arbeidssokerperioder;
    }

    @Override
    public String toString() {
        return "{" +
                "arbeidssokerperioder=" + arbeidssokerperioder +
                '}';
    }
}

File: src/main/kotlin/no/nav/fo/veilarbregistrering/arbeidssoker/Arbeidssokerperioder.kt
package no.nav.fo.veilarbregistrering.arbeidssoker

import no.nav.fo.veilarbregistrering.bruker.Periode
import java.util.Objects

class Arbeidssokerperioder(arbeidssokerperioder: List<Formidlingsgruppeperiode>?) {

    private val arbeidssokerperioder: List<Formidlingsgruppeperiode> = arbeidssokerperioder ?: emptyList()

    fun overlapperMed(forespurtPeriode: Periode): Arbeidssokerperioder {
        return Arbeidssokerperioder(arbeidssokerperioder
            .filter { it.periode.overlapperMed(forespurtPeriode) && it.formidlingsgruppe.erArbeidssoker() }
        )
    }

    fun dekkerHele(forespurtPeriode: Periode): Boolean {
        val eldsteFormidlingsgruppeperiode: Formidlingsgruppeperiode? = arbeidssokerperioder.minByOrNull { it.periode.fra }

        return eldsteFormidlingsgruppeperiode?.let { forespurtPeriode.fraOgMed(it.periode) } ?: false
    }

    fun asList(): List<Formidlingsgruppeperiode> {
        return arbeidssokerperioder
    }

    fun eldsteFoerst(): List<Formidlingsgruppeperiode> {
        return arbeidssokerperioder.sortedBy{ it.periode.fra }
    }

    override fun toString(): String {
        return "{arbeidssokerperioder=$arbeidssokerperioder}"
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val that = other as Arbeidssokerperioder
        return arbeidssokerperioder == that.arbeidssokerperioder
    }

    override fun hashCode(): Int {
        return Objects.hash(arbeidssokerperioder)
    }

    companion object {
        fun of(arbeidssokerperioder: List<Formidlingsgruppeperiode>?): Arbeidssokerperioder {
            return Arbeidssokerperioder(arbeidssokerperioder)
        }
    }

}